### PR TITLE
Swift: Models for String methods involving closures.

### PR DESCRIPTION
**Pull request by geoffw0**
Add flow models for [`String`](https://developer.apple.com/documentation/swift/string) methods involving closures, such as `String.withUTF8(_:)`.  Most already had test coverage, but I did have to add tests for `withMutableCharacters(_:)` and an additional test for `String(unsafeUninitializedCapacity:initializingUTF8With:)` to show that the model is working despite other issues in the existing test.  Unfortunately the diff of the test is also polluted by a renumbering.

### CODE REVIEW

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll [hunk 1]
```diff
@@ -68,6 +68,10 @@ private class StringSummaries extends SummaryModelCsv {
         ";StringProtocol;true;trimmingCharacters(in:);;;Argument[-1];ReturnValue;taint",
         ";StringProtocol;true;uppercased();;;Argument[-1];ReturnValue;taint",
         ";StringProtocol;true;uppercased(with:);;;Argument[-1];ReturnValue;taint",
+        ";StringProtocol;true;withCString(_:);;;Argument[-1];Argument[0].Parameter[0].CollectionElement;taint",
+        ";StringProtocol;true;withCString(_:);;;Argument[0].ReturnValue;ReturnValue;value",
```
Comment (rdmarsh2) [inline swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll:72]: Should there be a step from the mutated parameter of the callback to the `this` parameter of the method?

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll [hunk 2]
```diff
@@ -125,6 +130,15 @@ private class StringSummaries extends SummaryModelCsv {
         ";String;true;encode(to:);;;Argument[-1];Argument[0];taint",
         ";String;true;decodeCString(_:as:repairingInvalidCodeUnits:);;;Argument[0];ReturnValue.TupleElement[0];taint",
         ";String;true;decodeCString(_:as:repairingInvalidCodeUnits:);;;Argument[0].CollectionElement;ReturnValue.TupleElement[0];taint",
+        ";String;true;withUTF8(_:);;;Argument[-1];Argument[0].Parameter[0].CollectionElement;taint",
+        ";String;true;withUTF8(_:);;;Argument[0].Parameter[0].CollectionElement;Argument[-1];taint",
+        ";String;true;withUTF8(_:);;;Argument[0].ReturnValue;ReturnValue;value",
+        ";String;true;withPlatformString(_:);;;Argument[-1];Argument[0].Parameter[0].CollectionElement;taint",
+        ";String;true;withPlatformString(_:);;;Argument[0].ReturnValue;ReturnValue;value",
```
Comment (rdmarsh2) [inline swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll:137]: Same question here.

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll:137]: I don't think you can mutate the string with these versions.  I'll check...

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll:137]: Sorry for taking a while to get back to this.  I think it's only possible to modify the object using the methods marked `mutating`, that is `withMutableCharacters` and `withUTF8` in this PR (and I'm not sure it's practically possible with the latter, see the [doc](https://developer.apple.com/documentation/swift/string/withutf8(_:))).

Some experimentation [here](https://swiftfiddle.com/?c=H4sIAAAAAAAAA82VUWvCMBDHv8qRp1akVedgK3MPEwQfBoLzadlDpqmG2bQkV3WI332JabfqpkzGoA%2Bl4Z%2FL%2Fe%2Fu19AtmZKIiCRLFcIgzeWMoUgllYU0ftfIkxGbvrE5p0ZfMQUaIhijEnIOPaCEMUaJ3cuMhB4l2srU0z4lvtV1sBa46Lsj3pZKgAwVCGlXS46Q5Mhel3xkxB5MpGYxfyyk1KTk6q7fXzB179lANEkim8DfJyo89ymfWy%2FO2q2dfzXmy%2Bgz9EAqK65GBnlmhsI9xTNemLc7rSZMzbQwgqsjCwAXD16l6bKSY%2F%2B2M%2FuPJnb%2BOSRhCF6uOYzXIkboBp2g65egRkuGcaqSP%2FEa2leaBWWymvFr%2F5Lf4Sxqh9EBmzwNbi7D9JDHMVclrImBdVMjOJ2zSGy3tQPxw30C890ALoSGVHITFDYcrQKBvRBsasavj9Gdrt48gUamcChnfFMZqN0QUnNlz20oaQLDPcZK%2BNmhfiuqkvuyck7PqRFSSZpkZX4418Et2X0AY3tHiHwGAAA%3D) - notice that the changes are never reflected back in `s`.

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll (hunk 1 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll:72]: Discussed on the other comment.

**Review by rdmarsh2 — APPROVED** [review]

**Pull request merged**